core/distributed/vector.cpp

```diff
@@ -0,0 +1,392 @@
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/distributed/vector.hpp>
+
+
+#include "core/distributed/vector_kernels.hpp"
+#include "core/matrix/dense_kernels.hpp"
+
+
+namespace gko {
+namespace distributed {
+namespace vector {
+
+
+GKO_REGISTER_OPERATION(compute_norm2_sqr, dense::compute_norm2_sqr);
+GKO_REGISTER_OPERATION(compute_sqrt, dense::compute_sqrt);
+GKO_REGISTER_OPERATION(outplace_absolute_dense, dense::outplace_absolute_dense);
+GKO_REGISTER_OPERATION(build_local, distributed_vector::build_local);
+
+
+}  // namespace vector
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void Vector<ValueType, LocalIndexType, GlobalIndexType>::apply_impl(
+    const LinOp* b, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void Vector<ValueType, LocalIndexType, GlobalIndexType>::apply_impl(
+    const LinOp* alpha, const LinOp* b, const LinOp* beta, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Vector<ValueType, LocalIndexType, GlobalIndexType>::Vector(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm,
+    std::shared_ptr<const Partition<LocalIndexType, GlobalIndexType>> partition,
+    dim<2> global_size, dim<2> local_size, size_type stride)
+    : EnableLinOp<
+          Vector<ValueType, LocalIndexType, GlobalIndexType>>{exec,
+                                                              global_size},
+      DistributedBase{comm},
+      partition_{
+          partition
+              ? std::move(partition)
+              : gko::share(
+                    Partition<LocalIndexType, GlobalIndexType>::create(exec))},
```

Comment:
I am not sure if this is what we want. Storing an empty partition gives us no benefits and is probably more error-prone as users and devs might assume it is a valid object. I would probably prefer to throw if the partition is invalid.

Reply:
I would put validation checking off, until the PR for that is done. Until then, I would just assume that we have a valid partition. Still, I will change it to just use `std::move(partition)`.

Reply:
Yeah, I understand that initially you want to create the vector with a Partition object, but when going through the `read` path of filling in the vector, the partition object there takes preference. 

Another issue, which also probably the validation should take into account is that the local_size should match what the partition gives. We should also probably discuss what parameter ordering we want in this constructor and which ones should have default values and which not.

Reply:
I'm open to that discussion. I think `comm` since that has always needs to be defined for a distributed object. For me the partition is the next most important thing (if we keep it), since it describes how the object is distributed. The rest is sorted as usual. The only part that I could see changing, is the position of the partition, perhaps it makes sense to switch it with the global (and local?) size.